distributor/pkg/utils/utils.go

```diff
@@ -21,6 +21,30 @@ type EventMatcher struct {
 	Service string
 }
 
+// ExecutionContext is used for synchronizing components e.g.
+// to enable synchronization for smooth graceful shutdown
+type ExecutionContext struct {
+	context.Context
+	// Wg is the go wait group used for synchronization
+	// via this ExecutionContext
+	Wg *sync.WaitGroup
+
+	// CancelFn can be called whenever the components
+	// using this ExecutionContext shall exit/cancel
+	// Calling CancelFn will effectively cancel the underlying
+	// go context
+	CancelFn context.CancelFunc
+}
+
+// NewExecutionContext creates a new ExecutionContext with the given underlying go context.
+// waitGroupCount determines the number of components this ExecutionContext
+// is used in
+func NewExecutionContext(ctx context.Context, waitGroupCount int) *ExecutionContext {
+	wg := new(sync.WaitGroup)
+	wg.Add(waitGroupCount)
+	return &ExecutionContext{ctx, wg, func() {}}
```

Comment:
I would add to the `CancelFn` a log message at info level that it has been called

Reply:
i think there is no way to do that